In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [36]:
raw_data = pd.read_csv('wines_SPA.csv')
raw_data

winery             wine  year  rating  num_reviews  \
0            Teso La Monja            Tinto  2013     4.9           58   
1                   Artadi    Vina El Pison  2018     4.9           31   
2             Vega Sicilia            Unico  2009     4.8         1793   
3             Vega Sicilia            Unico  1999     4.8         1705   
4             Vega Sicilia            Unico  1996     4.8         1309   
...                    ...              ...   ...     ...          ...   
7495               Contino          Reserva  2016     4.2          392   
7496  Conreria d'Scala Dei    Les Brugueres  2018     4.2          390   
7497           Mustiguillo  Finca Terrerazo  2017     4.2          390   
7498           Matarromera     Gran Reserva  2011     4.2          389   
7499              Sei Solo         Preludio  2016     4.2          388   

     country            region   price                  type  body  acidity  
0     Espana              Toro  995.00              Toro Red   5.0      3.0  
1     Espana    Vino de Espana  313.50           Tempranillo   4.0      2.0  
2     Espana  Ribera del Duero  324.95  Ribera Del Duero Red   5.0      3.0  
3     Espana  Ribera del Duero  692.96  Ribera Del Duero Red   5.0      3.0  
4     Espana  Ribera del Duero  778.06  Ribera Del Duero Red   5.0      3.0  
...      ...               ...     ...                   ...   ...      ...  
7495  Espana             Rioja   19.98             Rioja Red   4.0      3.0  
7496  Espana          Priorato   16.76           Priorat Red   4.0      3.0  
7497  Espana      El Terrerazo   24.45                   Red   4.0      3.0  
7498  Espana  Ribera del Duero   64.50  Ribera Del Duero Red   5.0      3.0  
7499  Espana  Ribera del Duero   31.63  Ribera Del Duero Red   5.0      3.0  

[7500 rows x 11 columns]

## PREPROCESAMIENTO

In [37]:
df = raw_data.copy()

df.isnull().any()

winery         False
wine           False
year            True
rating         False
num_reviews    False
country        False
region         False
price          False
type            True
body            True
acidity         True
dtype: bool

In [38]:
df.dtypes

winery          object
wine            object
year            object
rating         float64
num_reviews      int64
country         object
region          object
price          float64
type            object
body           float64
acidity        float64
dtype: object

In [39]:
df.describe()

rating   num_reviews        price         body      acidity
count  7500.000000   7500.000000  7500.000000  6331.000000  6331.000000
mean      4.254933    451.109067    60.095822     4.158427     2.946612
std       0.118029    723.001856   150.356676     0.583352     0.248202
min       4.200000     25.000000     4.990000     2.000000     1.000000
25%       4.200000    389.000000    18.900000     4.000000     3.000000
50%       4.200000    404.000000    28.530000     4.000000     3.000000
75%       4.200000    415.000000    51.350000     5.000000     3.000000
max       4.900000  32624.000000  3119.080000     5.000000     3.000000

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   winery       7500 non-null   object 
 1   wine         7500 non-null   object 
 2   year         7498 non-null   object 
 3   rating       7500 non-null   float64
 4   num_reviews  7500 non-null   int64  
 5   country      7500 non-null   object 
 6   region       7500 non-null   object 
 7   price        7500 non-null   float64
 8   type         6955 non-null   object 
 9   body         6331 non-null   float64
 10  acidity      6331 non-null   float64
dtypes: float64(4), int64(1), object(6)
memory usage: 644.7+ KB


REVISAMOS LA CATEGORIA YEAR

In [46]:
#Revisamos la cateogoria "year"
df.year.unique()

array(['2013', '2018', '2009', '1999', '1996', '1998', '2010', '1995',
       '2015', '2011', '2016', '1970', '1946', '1962', '2019', '2004',
       'N.V.', '1931', '1979', '2005', '2020', '2014', '1985', '1929',
       '2007', '2012', '2017', '2008', nan, '2006', '2000', '2003',
       '2002', '1991', '1994', '1990', '1989', '1987', '1986', '1981',
       '2001', '1968', '1964', '1982', '1974', '1983', '1955', '1980',
       '1972', '1953', '1958', '1942', '1965', '1992', '1973', '1997',
       '1967', '1975', '1910', '1961', '1954', '1988', '1969', '1951',
       '1928', '1976', '1949', '2021', '1959', '1922', '1978', '1925'],
      dtype=object)

In [48]:
#Vemos que hay datos con la leyenda N.V. y otros nulos
df['year'].value_counts()['N.V.']

288

In [49]:
df['year'].isna().value_counts()

year
False    7498
True        2
Name: count, dtype: int64

In [67]:
#como hay dos registros se procede a eliminarlos
df.dropna(subset=['year'], inplace=True)

REVISAMOS LAS CATEGORIAS BODY Y ADICITY, YA QUE TIENEN LA MISMA CANTIDAD DE NULLS

In [62]:
#Como son muchos registros, vamos a proceder a imputarle a los datos faltantes la media.
df['body'] = df['body'].fillna(df['body'].mean())
df['body'].isna().value_counts()

body
False    7500
Name: count, dtype: int64

In [63]:
df['acidity'] = df['acidity'].fillna(df['acidity'].mean())
df['acidity'].isna().value_counts()

acidity
False    7500
Name: count, dtype: int64

REVISAMOS LA CATEGORIA TYPE

In [65]:
#Como es un dato categorico, vamos a reemplazarlo por el valor mas común.
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(missing_values=np.NaN, strategy="most_frequent")
df.type = imputer.fit_transform(df.type.values.reshape(-1,1))[:,0]

In [68]:
#VERIFICAMOS LOS VALORES NULL
df.isna().sum()

winery         0
wine           0
year           0
rating         0
num_reviews    0
country        0
region         0
price          0
type           0
body           0
acidity        0
dtype: int64

In [69]:
df.duplicated().sum()

5452

SOLO VAMOS A USAR LAS CATEGORIAS RATING, PRICE, TYPE, BODY, ACIDITY

In [70]:
df = df.drop(['winery','wine','year','num_reviews','country','region'], axis=1)
df

rating   price                  type  body  acidity
0        4.9  995.00              Toro Red   5.0      3.0
1        4.9  313.50           Tempranillo   4.0      2.0
2        4.8  324.95  Ribera Del Duero Red   5.0      3.0
3        4.8  692.96  Ribera Del Duero Red   5.0      3.0
4        4.8  778.06  Ribera Del Duero Red   5.0      3.0
...      ...     ...                   ...   ...      ...
7495     4.2   19.98             Rioja Red   4.0      3.0
7496     4.2   16.76           Priorat Red   4.0      3.0
7497     4.2   24.45                   Red   4.0      3.0
7498     4.2   64.50  Ribera Del Duero Red   5.0      3.0
7499     4.2   31.63  Ribera Del Duero Red   5.0      3.0

[7498 rows x 5 columns]

In [71]:
df.duplicated().sum()

5549

In [76]:
#VAMOS A ELIMINAR LOS DUPLICADOS

df = df.drop_duplicates()
df.duplicated().sum()
df.dropna(axis=0, inplace=True)

In [77]:
df

rating   price                  type      body   acidity
0        4.9  995.00              Toro Red  5.000000  3.000000
1        4.9  313.50           Tempranillo  4.000000  2.000000
2        4.8  324.95  Ribera Del Duero Red  5.000000  3.000000
3        4.8  692.96  Ribera Del Duero Red  5.000000  3.000000
4        4.8  778.06  Ribera Del Duero Red  5.000000  3.000000
...      ...     ...                   ...       ...       ...
2046     4.2   24.45                   Red  4.000000  3.000000
2047     4.2   64.50  Ribera Del Duero Red  5.000000  3.000000
2048     4.2   31.63  Ribera Del Duero Red  5.000000  3.000000
2049     4.2   73.00             Rioja Red  4.000000  3.000000
6100     4.2   29.15           Rioja White  4.158427  2.946612

[1949 rows x 5 columns]

APLICACION ONE HOT A TYPE

In [79]:
df_final = df.copy()
df_final = pd.get_dummies(df_final, columns=['type']).astype(np.int64)
df_final

rating  price  body  acidity  type_Albarino  type_Cabernet Sauvignon  \
0          4    995     5        3              0                        0   
1          4    313     4        2              0                        0   
2          4    324     5        3              0                        0   
3          4    692     5        3              0                        0   
4          4    778     5        3              0                        0   
...      ...    ...   ...      ...            ...                      ...   
2046       4     24     4        3              0                        0   
2047       4     64     5        3              0                        0   
2048       4     31     5        3              0                        0   
2049       4     73     4        3              0                        0   
6100       4     29     4        2              0                        0   

      type_Cava  type_Chardonnay  type_Grenache  type_Mencia  ...  \
0             0                0              0            0  ...   
1             0                0              0            0  ...   
2             0                0              0            0  ...   
3             0                0              0            0  ...   
4             0                0              0            0  ...   
...         ...              ...            ...          ...  ...   
2046          0                0              0            0  ...   
2047          0                0              0            0  ...   
2048          0                0              0            0  ...   
2049          0                0              0            0  ...   
6100          0                0              0            0  ...   

      type_Ribera Del Duero Red  type_Rioja Red  type_Rioja White  \
0                             0               0                 0   
1                             0               0                 0   
2                             1               0                 0   
3                             1               0                 0   
4                             1               0                 0   
...                         ...             ...               ...   
2046                          0               0                 0   
2047                          1               0                 0   
2048                          1               0                 0   
2049                          0               1                 0   
6100                          0               0                 1   

      type_Sauvignon Blanc  type_Sherry  type_Sparkling  type_Syrah  \
0                        0            0               0           0   
1                        0            0               0           0   
2                        0            0               0           0   
3                        0            0               0           0   
4                        0            0               0           0   
...                    ...          ...             ...         ...   
2046                     0            0               0           0   
2047                     0            0               0           0   
2048                     0            0               0           0   
2049                     0            0               0           0   
6100                     0            0               0           0   

      type_Tempranillo  type_Toro Red  type_Verdejo  
0                    0              1             0  
1                    1              0             0  
2                    0              0             0  
3                    0              0             0  
4                    0              0             0  
...                ...            ...           ...  
2046                 0              0             0  
2047                 0              0             0  
2048                 0              0             0  
2049              

## ESTANDARIZACION / DIVISION DE DATOS PARA ENTRENAMIENTO Y PRUEBA

In [88]:
from sklearn.model_selection import train_test_split
from sklearn import preprocessing

df_scaled = preprocessing.scale(df_final)
df_scaled = pd.DataFrame(df_scaled, columns=df_final.columns)
df_scaled




rating     price      body   acidity  type_Albarino  \
0        0.0  3.208458  1.254378  0.461225      -0.133246   
1        0.0  0.665502 -0.361473 -1.786094      -0.133246   
2        0.0  0.706518  1.254378  0.461225      -0.133246   
3        0.0  2.078670  1.254378  0.461225      -0.133246   
4        0.0  2.399336  1.254378  0.461225      -0.133246   
...      ...       ...       ...       ...            ...   
1944     0.0 -0.412084 -0.361473  0.461225      -0.133246   
1945     0.0 -0.262938  1.254378  0.461225      -0.133246   
1946     0.0 -0.385984  1.254378  0.461225      -0.133246   
1947     0.0 -0.229379 -0.361473  0.461225      -0.133246   
1948     0.0 -0.393441 -0.361473 -1.786094      -0.133246   

      type_Cabernet Sauvignon  type_Cava  type_Chardonnay  type_Grenache  \
0                   -0.071814    -0.1292        -0.078709      -0.131238   
1                   -0.071814    -0.1292        -0.078709      -0.131238   
2                   -0.071814    -0.1292        -0.078709      -0.131238   
3                   -0.071814    -0.1292        -0.078709      -0.131238   
4                   -0.071814    -0.1292        -0.078709      -0.131238   
...                       ...        ...              ...            ...   
1944                -0.071814    -0.1292        -0.078709      -0.131238   
1945                -0.071814    -0.1292        -0.078709      -0.131238   
1946                -0.071814    -0.1292        -0.078709      -0.131238   
1947                -0.071814    -0.1292        -0.078709      -0.131238   
1948                -0.071814    -0.1292        -0.078709      -0.131238   

      type_Mencia  ...  type_Ribera Del Duero Red  type_Rioja Red  \
0       -0.093804  ...                  -0.594535       -0.601652   
1       -0.093804  ...                  -0.594535       -0.601652   
2       -0.093804  ...                   1.681986       -0.601652   
3       -0.093804  ...                   1.681986       -0.601652   
4       -0.093804  ...                   1.681986       -0.601652   
...           ...  ...                        ...             ...   
1944    -0.093804  ...                  -0.594535       -0.601652   
1945    -0.093804  ...                   1.681986       -0.601652   
1946    -0.093804  ...                   1.681986       -0.601652   
1947    -0.093804  ...                  -0.594535        1.662092   
1948    -0.093804  ...                  -0.594535       -0.601652   

      type_Rioja White  type_Sauvignon Blanc  type_Sherry  type_Sparkling  \
0            -0.137181             -0.045349    -0.170409       -0.050715   
1            -0.137181             -0.045349    -0.170409       -0.050715   
2            -0.137181             -0.045349    -0.170409       -0.050715   
3            -0.137181             -0.045349    -0.170409       -0.050715   
4            -0.137181             -0.045349    -0.170409       -0.050715   
...                ...                   ...          ...             ...   
1944         -0.137181             -0.045349    -0.170409       -0.050715   
1945         -0.137181             -0.045349    -0.170409       -0.050715   
1946         -0.137181             -0.045349    -0.170409       -0.050715   
1947         -0.137181             -0.045349    -0.170409       -0.050715   
1948          7.289643             -0.045349    -0.170409       -0.050715   

      type_Syrah  type_Tempranillo  type_Toro Red  type_Verdejo  
0      -0.088068         -0.195855       4.897671     -0.118524  
1      -0.088068          5.105825      -0.204179     -0.118524  
2      -0.088068         -0.195855      -0.204179     -0.118524  
3      -0.088068         -0.195855      -0.204179     -0.118524  
4      -0.088068         -0.195855      -0.204179     -0.118524  
...          ...               ...            ...           ...  
1944   -0.088068         -0.195855      -0.204179     -0.118524  
1945   -0.088068         -0.195855      -0.204179     -0.118524  
1946   -0.08

In [90]:
X = df_scaled.drop(['price'],axis=1)
Y = df_scaled['price']

X_train, X_valid, Y_train, Y_valid = train_test_split(X,Y,train_size=0.8,test_size=0.2,random_state=0)

## MODELOS DE PRUEBA

In [91]:
from sklearn.metrics import mean_absolute_percentage_error
from sklearn import svm
from sklearn.svm import SVC

model_SVR = svm.SVR()
model_SVR.fit(X_train, Y_train)
Y_pred = model_SVR.predict(X_valid)
mean_absolute_percentage_error(Y_valid, Y_pred)

1.6863536012426867

In [92]:
from sklearn.ensemble import RandomForestRegressor

model_RFR = RandomForestRegressor(n_estimators=10)
model_RFR.fit(X_train, Y_train)
Y_pred = model_RFR.predict(X_valid)
mean_absolute_percentage_error(Y_valid, Y_pred)

2.3391820702470225

In [93]:
from sklearn.linear_model import LinearRegression

model_LR = LinearRegression()
model_LR.fit(X_train, Y_train)
Y_pred = model_LR.predict(X_valid)
mean_absolute_percentage_error(Y_valid, Y_pred)

2.2767796577509776

Luego del preprocesamiento, estandarización de datos y elección de categorias relevantes, procedimos a probar 3 modelos, el cual SVR resulto con el mejor valor predictivo con un % de error de 1.68.